### PR TITLE
`build_optimizer()` assign all parameters

### PR DESCRIPTION
**Pull request by glenn-jocher**
@AyushExel @Laughing-q should resolve un-assigned parameters in optimizer.

### Summary
🛠️📖🚀

Refactored `trainer.py` to use `named_modules` and `named_parameters` for better readability and robustness. Added `fullname` variable to each parameter for logging or debugging.

> _`named_modules` used_
> _refactoring for clarity_
> _autumn leaves falling_

### Walkthrough
*  Refactor code to use `named_modules` and `named_parameters` instead of `modules` and `hasattr` ([link](https://github.com/ultralytics/ultralytics/pull/2855/files?diff=unified&w=0#diff-1dc4dec7a1716e080109cc732dc44c6e843b30bcdd324e0141e679ebf718c2b1L621-R634))
*  Assign a `fullname` variable to each parameter for logging or debugging purposes ([link](https://github.com/ultralytics/ultralytics/pull/2855/files?diff=unified&w=0#diff-1dc4dec7a1716e080109cc732dc44c6e843b30bcdd324e0141e679ebf718c2b1L621-R634))
*  Improve code quality and performance of the YOLOv5 model in `ultralytics/yolo/engine/trainer.py` ([link](https://github.com/ultralytics/ultralytics/pull/2855/files?diff=unified&w=0#diff-1dc4dec7a1716e080109cc732dc44c6e843b30bcdd324e0141e679ebf718c2b1L621-R634))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced optimizer parameter grouping for better model training in Ultralytics' YOLO.

### 📊 Key Changes
- Parameter grouping logic in `build_optimizer` has been refined.
- Module and parameter names are now fully qualified during the grouping process, improving clarity.
- Bias, normalization layer weights, and other weights are more accurately categorized for optimizer configuration.

### 🎯 Purpose & Impact
- 🎨 **Clarity**: The clear naming of parameters helps developers understand exactly which parameters are being grouped for optimization.
- 💪 **Efficiency**: Better parameter grouping can lead to more efficient training by correctly applying weight decay only where appropriate.
- 🚀 **Performance**: Improved optimizer setup can potentially enhance model training performance and accuracy.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
### YOLOv8n
```
# main
optimizer: SGD(lr=0.01) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias

# this PR
optimizer: SGD(lr=0.01) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias
```

**Comment by glenn-jocher** [comment]
### RTDETR-l

```
### rtdetr_loss branch
optimizer: SGD(lr=0.01) with parameter groups 143 weight(decay=0.0), 199 weight(decay=0.0005), 219 bias

### this fix applied to rtdetr_loss branch
optimizer: SGD(lr=0.01) with parameter groups 143 weight(decay=0.0), 206 weight(decay=0.0005), 226 bias
```

**Comment by AyushExel** [comment]
@glenn-jocher if you merge this to main now, we can pull the changes to rtdetr_loss branch and test on a 2 free GPUs

**Comment by glenn-jocher** [comment]
@AyushExel ya I'll do that once the tests finish!

**Comment by Laughing-q** [comment]
@glenn-jocher @AyushExel Just checked the params, I think you guys might be interested in this. The missing params in optimizer is `in_proj_weight` and `out_proj.weight` in `nn.MultiheadAttention`. :)
